File: terrain/tests/parity/java/DensityParityTest.java
import java.io.BufferedOutputStream;
import java.io.DataOutputStream;
import java.io.FileOutputStream;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;
import net.minecraft.SharedConstants;
import net.minecraft.core.HolderGetter;
import net.minecraft.core.HolderLookup;
import net.minecraft.core.LayeredRegistryAccess;
import net.minecraft.core.Registry;
import net.minecraft.core.RegistryAccess;
import net.minecraft.core.registries.Registries;
import net.minecraft.resources.Identifier;
import net.minecraft.resources.RegistryDataLoader;
import net.minecraft.resources.ResourceKey;
import net.minecraft.server.Bootstrap;
import net.minecraft.server.RegistryLayer;
import net.minecraft.server.packs.repository.ServerPacksSource;
import net.minecraft.server.packs.PackType;
import net.minecraft.server.packs.resources.MultiPackResourceManager;
import net.minecraft.server.packs.resources.ResourceManager;
import net.minecraft.tags.TagLoader;
import net.minecraft.world.level.levelgen.NoiseGeneratorSettings;
import net.minecraft.world.level.levelgen.NoiseRouter;
import net.minecraft.world.level.levelgen.RandomState;
import net.minecraft.world.level.levelgen.densityfunction.DensityBuffer;
import net.minecraft.world.level.levelgen.densityfunction.DensityBufferPool;
import net.minecraft.world.level.levelgen.densityfunction.DensityFunction;
import net.minecraft.world.level.levelgen.densityfunction.DensitySamplerSet;
import net.minecraft.world.level.levelgen.densityfunction.DensityVolume;
import net.minecraft.world.level.levelgen.densityfunction.SamplerContext;
import net.minecraft.world.level.levelgen.densityfunction.ScopedDensityBuffer;
import net.minecraft.world.level.levelgen.synth.NormalNoise;

/**
 * Density engine parity: every noise_settings router (and aquifer) function,
 * decoded from the vanilla datapack exactly as a world load does, sampled
 * three ways - uncached points, cached chunk and quart volumes, and cached
 * points answered from those volumes. The C++ twin (cpp/DensityParityTest.cpp)
 * writes the same stream; compare with compare_density_parity.py.
 *
 * Usage: DensityParityTest <seed> <out.txt> <out.bin>
 */
public class DensityParityTest {
    static final String[] SETTINGS = {"overworld", "large_biomes", "amplified", "nether", "end", "caves", "floating_islands"};
    static final int[][] CHUNKS = {{0, 0}, {-7, 12}, {123, -456}, {-3000, 2500}};

    public static void main(String[] args) throws Exception {
        SharedConstants.tryDetectVersion();
        Bootstrap.bootStrap();
        long seed = Long.parseLong(args[0]);

        // The vanilla datapack (server data), as WorldLoader opens it.
        ResourceManager resources = new MultiPackResourceManager(PackType.SERVER_DATA,
            List.of(ServerPacksSource.createVanillaPackSource().fullResources()));
        LayeredRegistryAccess<RegistryLayer> layers = RegistryLayer.createRegistryAccess();
        List<Registry.PendingTags<?>> staticTags = TagLoader.loadTagsForExistingRegistries(resources, layers.getLayer(RegistryLayer.STATIC));
        RegistryAccess.Frozen context = layers.getAccessForLoading(RegistryLayer.WORLD);
        List<HolderLookup.RegistryLookup<?>> contextRegistries = TagLoader.buildUpdatedLookups(context, staticTags);
        RegistryAccess.Frozen world = RegistryDataLoader.load(resources, contextRegistries, RegistryDataLoader.WORLD_REGISTRIES, Runnable::run).join();
        HolderGetter<NormalNoise> noises = world.lookupOrThrow(Registries.NOISE);

        try (PrintWriter text = new PrintWriter(args[1]);
             DataOutputStream bin = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(args[2])))) {
            for (String name : SETTINGS) {
                NoiseGeneratorSettings settings = world.lookupOrThrow(Registries.NOISE_SETTINGS)
                    .getOrThrow(ResourceKey.create(Registries.NOISE_SETTINGS, Identifier.withDefaultNamespace(name))).value();
                RandomState randomState = RandomState.create(noises, seed, settings);
                List<String> labels = new ArrayList<>();
                List<DensityFunction> functions = new ArrayList<>();
                NoiseRouter r = settings.noiseRouter();
                labels.add("temperature"); functions.add(r.temperature());
                labels.add("vegetation"); functions.add(r.vegetation());
                labels.add("continents"); functions.add(r.continents());
                labels.add("erosion"); functions.add(r.erosion());
                labels.add("depth"); functions.add(r.depth());
                labels.add("ridges"); functions.add(r.ridges());
                labels.add("chunk_surface_level"); functions.add(r.chunkSurfaceLevel());
                labels.add("final_density"); functions.add(r.finalDensity());
                settings.aquifers().ifPresent(a -> {
                    labels.add("aq_barrier"); functions.add(a.barrierNoise());
                    labels.add("aq_floodedness"); functions.add(a.fluidLevelFloodednessNoise());
                    labels.add("aq_spread"); functions.add(a.fluidLevelSpreadNoise());
                    labels.add("aq_lava"); functions.add(a.lavaNoise());
                    labels.add("aq_exclusion"); functions.add(a.exclusion());
                    labels.add("aq_surface_level"); functions.add(a.surfaceLevel());
                });
                int minY = settings.noiseSettings().minY();
                int height = settings.noiseSettings().height();

                // Uncached points.
                for (int f = 0; f < functions.size(); f++) {
                    for (int i = 0; i < 64; i++) {
                        int x = pointX(i), y = minY + Math.floorMod(i * 37, height), z = pointZ(i);
                        float v = randomState.sampleBlockValueUncached(functions.get(f), x, y, z);
                        text.println("P " + name + " " + labels.get(f) + " " + x + " " + y + " " + z + " " + hex(v));
                    }
                }
                // Cached volumes, then cached points inside them, one context per chunk.
                for (int[] c : CHUNKS) {
                    DensityBufferPool pool = randomState.acquireDensityBufferPool();
                    SamplerContext sc = SamplerContext.builder().useBufferArena(pool).enableCaches().build();
                    DensitySamplerSet samplers = randomState.samplersWithContext(sc);
                    DensityVolume chunk = new DensityVolume(16, height, 16, c[0] * 16, minY, c[1] * 16);
                    DensityVolume quart = new DensityVolume(4, height / 4, 4, c[0] * 16, minY, c[1] * 16, 4, 4, 4);
                    for (int f = 0; f < functions.size(); f++) {
                        for (DensityVolume v : new DensityVolume[]{chunk, quart}) {
                            try (ScopedDensityBuffer b = samplers.get(functions.get(f)).sampleVolume(v)) {
                                text.println("V " + name + " " + labels.get(f) + " " + c[0] + " " + c[1] + " " + v.sizeX() + "x" + v.sizeY() + "x" + v.sizeZ() + " " + v.size());
                                for (int i = 0; i < v.size(); i++) bin.writeInt(Float.floatToRawIntBits(b.get(i)));
                            }
                        }
                        for (int i = 0; i < 16; i++) {
                            int x = c[0] * 16 + (i * 5) % 16, y = minY + Math.floorMod(i * 53, height), z = c[1] * 16 + (i * 11) % 16;
                            float v = samplers.sampleValue(functions.get(f), x, y, z);
                            text.println("C " + name + " " + labels.get(f) + " " + x + " " + y + " " + z + " " + hex(v));
                        }
                    }
                    randomState.releaseDensityBufferPool(pool);
                }
            }
        }
    }

    static int pointX(int i) { return (int) ((i * 2654435761L) % 60001L) - 30000; }
    static int pointZ(int i) { return (int) ((i * 40503L + 17L) % 60001L) - 30000; }
    static String hex(float v) { return Integer.toHexString(Float.floatToRawIntBits(v)); }
}

File: terrain/tests/parity/cpp/DensityParityTest.cpp
// Density engine parity, C++ side of java/DensityParityTest.java: every
// noise_settings router (and aquifer) function decoded from the datapack JSON
// (data/minecraft/worldgen) through WorldgenRegistries, sampled as uncached
// points, cached chunk and quart volumes, and cached points answered from
// those volumes. Writes the same text + binary stream as the Java test;
// compare_density_parity.py diffs the two.
//
// Usage: density_parity_test <seed> <out.txt> <out.bin>   (run from the repo,
// or with MC_DATA_ROOT pointing at data/)

#include "levelgen/density/DensityBuffer.h"
#include "levelgen/density/DensityFunctionCompiler.h"
#include "levelgen/density/DensityVolume.h"
#include "levelgen/density/SamplerContext.h"
#include "levelgen/density/WorldgenRegistries.h"
#include "levelgen/density/terrain/RandomState.h"
#include "levelgen/density/terrain/TerrainSettings.h"
#include "world/level/block/Blocks.h"

#include <cstdint>
#include <cstdio>
#include <cstring>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

using namespace minecraft::levelgen::density;

namespace {

const char* const kSettings[] = {"overworld", "large_biomes", "amplified", "nether", "end", "caves", "floating_islands"};
const int kChunks[][2] = {{0, 0}, {-7, 12}, {123, -456}, {-3000, 2500}};

int floorMod(int a, int b) {
    const int m = a % b;
    return m < 0 ? m + b : m;
}
int pointX(int i) { return static_cast<int>((i * 2654435761LL) % 60001LL) - 30000; }
int pointZ(int i) { return static_cast<int>((i * 40503LL + 17LL) % 60001LL) - 30000; }

std::string hex(float v) {
    uint32_t bits;
    std::memcpy(&bits, &v, sizeof(bits));
    char buffer[16];
    std::snprintf(buffer, sizeof(buffer), "%x", bits);
    return buffer;
}

void writeBigEndian(std::ofstream& out, float v) {
    uint32_t bits;
    std::memcpy(&bits, &v, sizeof(bits));
    const unsigned char bytes[4] = {static_cast<unsigned char>(bits >> 24), static_cast<unsigned char>(bits >> 16),
                                    static_cast<unsigned char>(bits >> 8), static_cast<unsigned char>(bits)};
    out.write(reinterpret_cast<const char*>(bytes), 4);
}

} // namespace

int main(int argc, char** argv) {
    if (argc < 4) {
        std::cerr << "usage: density_parity_test <seed> <out.txt> <out.bin>\n";
        return 2;
    }
    const int64_t seed = std::stoll(argv[1]);
    minecraft::world::level::block::Blocks::bootstrap();
    WorldgenRegistries& registries = WorldgenRegistries::get();

    std::ofstream text(argv[2]);
    std::ofstream bin(argv[3], std::ios::binary);
    try {
        for (const char* name : kSettings) {
            std::shared_ptr<const TerrainSettings> settings = TerrainSettings::load(name, registries);
            RandomState randomState(seed, settings->useLegacyRandomSource, registries);
            std::vector<std::string> labels;
            std::vector<DensityFunctionPtr> functions;
            const NoiseRouter& r = settings->noiseRouter;
            labels = {"temperature", "vegetation", "continents", "erosion", "depth", "ridges", "chunk_surface_level",
                      "final_density"};
            functions = {r.temperature, r.vegetation, r.continents, r.erosion, r.depth, r.ridges, r.chunkSurfaceLevel,
                         r.finalDensity};
            if (settings->aquifers) {
                const Aquifer::Config& a = *settings->aquifers;
                for (const char* label : {"aq_barrier", "aq_floodedness", "aq_spread", "aq_lava", "aq_exclusion",
                                          "aq_surface_level"}) {
                    labels.emplace_back(label);
                }
                for (const DensityFunctionPtr& f : {a.barrierNoise, a.fluidLevelFloodednessNoise, a.fluidLevelSpreadNoise,
                                                    a.lavaNoise, a.exclusion, a.surfaceLevel}) {
                    functions.push_back(f);
                }
            }
            const int minY = settings->noiseSettings.minY;
            const int height = settings->noiseSettings.height;

            for (size_t f = 0; f < functions.size(); ++f) {
                for (int i = 0; i < 64; ++i) {
                    const int x = pointX(i), y = minY + floorMod(i * 37, height), z = pointZ(i);
                    const float v = randomState.sampleBlockValueUncached(functions[f], x, y, z);
                    text << "P " << name << ' ' << labels[f] << ' ' << x << ' ' << y << ' ' << z << ' ' << hex(v) << '\n';
                }
            }
            for (const auto& c : kChunks) {
                std::unique_ptr<DensityBufferPool> pool = randomState.acquireDensityBufferPool();
                {
                    SamplerContext sc = SamplerContext::builder().useBufferArena(*pool).enableCaches().build();
                    DensitySamplerSet samplers = randomState.samplersWithContext(sc);
                    const DensityVolume chunk(16, height, 16, c[0] * 16, minY, c[1] * 16);
                    const DensityVolume quart(4, height / 4, 4, c[0] * 16, minY, c[1] * 16, 4, 4, 4);
                    for (size_t f = 0; f < functions.size(); ++f) {
                        for (const DensityVolume* v : {&chunk, &quart}) {
                            ScopedBuffer buffer = samplers.get(functions[f]).sampleVolume(*v);
                            text << "V " << name << ' ' << labels[f] << ' ' << c[0] << ' ' << c[1] << ' ' << v->sizeX
                                 << 'x' << v->sizeY << 'x' << v->sizeZ << ' ' << v->size() << '\n';
                            for (int i = 0; i < v->size(); ++i) writeBigEndian(bin, buffer->get(i));
                        }
                        for (int i = 0; i < 16; ++i) {
                            const int x = c[0] * 16 + (i * 5) % 16, y = minY + floorMod(i * 53, height),
                                      z = c[1] * 16 + (i * 11) % 16;
                            const float v = samplers.sampleValue(functions[f], x, y, z);
                            text << "C " << name << ' ' << labels[f] << ' ' << x << ' ' << y << ' ' << z << ' ' << hex(v)
                                 << '\n';
                        }
                    }
                }
                randomState.releaseDensityBufferPool(std::move(pool));
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "density_parity_test: " << e.what() << '\n';
        return 1;
    }
    return 0;
}
